---
title: "A Tutorial of Sample-Pseudotime Aanalysis on COVID scRNA-seq Data"
author: 
- name: Runzhe Li
  affiliation: Johns Hopkins University
  email: rli51@jhmi.edu
date: "`r Sys.Date()`"
output: 
  BiocStyle::html_document:
    toc_float: true
abstract: |
  Description of your vignette
vignette: |
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(message = F, warning = F)
```

# Load packages
```{r}
library(tidyverse)
library(TSCAN)
library(ggpubr)
library(RColorBrewer)
library(mgcv)
library(ComplexHeatmap)
```

# Load data
* `pbpca`: the top 20 PCs of pseudo-bulk gene expression data

* `cell_prop_pca`: the top 3 PCs of cell type proportion data

* `patient_info`: the patient-level metadata

* `sample_info_pb_center`: the patient-level metadata with predefined sample cluster info. 
```{r}
pbpca = readRDS("./data/pbpca.rds")
cell_prop_pca = readRDS("./data/cell_prop_pca.rds")
patient_info = readRDS("./data/patient_info.rds")
sample_info_pb_center = readRDS("./data/sample_info_pb_center.rds")

print(str(pbpca))
print(str(cell_prop_pca))
print(head(patient_info))
print(head(sample_info_pb_center))
```

# Construct Sample Pseudotime

To construct pseudotime, we provide the function `get_sample_pseudotime`:

* `x`: the data matrix of features $\times$ samples, i.e. each row is a feature of gene expression or cell type proportion, each column is a patient sample.

* `sample_metadata`: patient-level metadata, including several important variables.

* `sample_name` and `phenotype_name`: the variable name in `sample_metadata` that indicates the sample and phenotype. The default setting of `sample_name` is "Patient" and `phenotype_name` is "Phenotype". 

* `method`: users now have two options, either using TSCAN for sample-level pseudotime construction or use CCA for optimal severity axis detection. 

* `cluster = NULL, reduce = F, clusternum = 2, startcluster = 1`: parameters for TSCAN.

* `phenotype_num`: the numerical value to indicate the phenotype severity. 


```{r}
get_sample_pseudotime = function(x, sample_metadata = NULL, 
                                 sample_name = 'Patient', phenotype_name = 'Phenotype',
                                 method = c('tscan','cca'),
                                 cluster = NULL, reduce = F, clusternum = 2, startcluster = 1,
                                 phenotype_num = NULL){
  method = match.arg(method)
  
  ## match x and meta
  match_idx = match(colnames(x), sample_metadata[,sample_name])
  sample_metadata = sample_metadata[match_idx, ]
  
  ## tscan
  if (method == 'tscan'){
    sample_cluster = exprmclust(x, 
                                reduce = reduce,
                                cluster = cluster,
                                clusternum = clusternum)
    sample_metadata$cluster = sample_cluster$clusterid
    
    sample_order = TSCANorder(sample_cluster,
                              startcluster = startcluster,
                              orderonly = F,
                              listbranch = F)
    
    sample_info = sample_metadata %>%
      inner_join(sample_order, by = setNames("sample_name",sample_name)) %>%
      arrange(Pseudotime)
  }
  if (method == 'cca'){
    if (reduce) {
      pc = prcomp(t(x), scale = T)$x[1:min(20, nrow(x))]
    }
    else {
      pc = t(x)
    }
    res = cancor(pc,phenotype_num)
    proj = (res$xcoef["PC1",1]*(pc[,"PC1"]-res$xcenter["PC1"]))+ (res$xcoef["PC2",1]*(pc[,"PC1"]-res$xcenter["PC2"]))
    pseudotime = order(proj)
    slope = res$xcoef["PC2",1]/res$xcoef["PC1",1]
    print(paste("slope is", slope))
    
    sample_info = sample_metadata %>%
      mutate(Pseudotime = pseudotime) %>%
      arrange(Pseudotime)
  }
  
  return(sample_info)
}
```

# Examples
We will walk through several examples...

```{r, echo=F}
theme_paper = function(){
  theme_classic()+
    theme(plot.title = element_text(hjust = 0.5)) 
  
} 

Study.color = c(brewer.pal(5,'Set1'),brewer.pal(7,'Set2')); names(Study.color) = sort(unique(patient_info$Study))
Phenotype.color = c("HD" = '#4DAF4A', "Mi" = '#377EB8' ,"Mod" ='orange', "Se" = '#E41A1C',
                    "Rec" = 'grey', "Flu" = 'purple')

create_pseudotime_plot = function(sample_info, title, cluster = F, opt_axis = F){
  p1 =  sample_info %>%
    ggplot(aes(x = PC1, y = PC2, color = Phenotype))+
    geom_point(size = 3)+
    coord_fixed(1)+
    scale_colour_manual(values = Phenotype.color)+
    guides(colour = guide_legend(nrow=2,byrow=TRUE)) + 
    labs(color = "Severity") +
    ggtitle(title)+
    theme_paper()
  
  if(cluster){
    p1 =  sample_info %>%
    mutate(cluster = factor(cluster)) %>%
    ggplot(aes(x = PC1, y = PC2, color = Phenotype, alpha = cluster))+
    geom_point(size = 3)+
    coord_fixed(1)+
    scale_colour_manual(values = Phenotype.color)+
    guides(colour = guide_legend(nrow=2,byrow=TRUE)) + 
    labs(color = "Severity") +
    ggtitle(title)+
    theme_paper()
  }
  
  if(opt_axis){
    p1 =  sample_info %>%
    mutate(cluster = factor(cluster)) %>%
    ggplot(aes(x = PC1, y = PC2, color = Phenotype))+
    geom_point(size = 3)+
    geom_abline(intercept = 0, slope = -0.1542, linetype = 'dashed')+
    coord_fixed(1)+
    scale_colour_manual(values = Phenotype.color)+
    guides(colour = guide_legend(nrow=2,byrow=TRUE)) + 
    labs(color = "Severity") +
    ggtitle(title)+
    theme_paper()
  }
  
  print(p1)
}
```


## Example 1: pseudobulk pca + tscan + model-based clustering with 2 clusters
```{r}
sample_info1 = get_sample_pseudotime(x = t(pbpca[,1:2]), sample_metadata = patient_info, 
                                    sample_name = 'Patient', phenotype_name = 'Phenotype',
                                    method = 'tscan',
                                    reduce = F, clusternum = 2, startcluster = 1)
sample_info_pc1 = pbpca %>% 
  as.data.frame() %>%
  select(PC1:PC2) %>%
  mutate(Patient = rownames(.)) %>%
  inner_join(sample_info1 %>% select(Patient, Phenotype, cluster, Pseudotime), by = 'Patient')

create_pseudotime_plot(sample_info_pc1, "Gene Expression", cluster = T)

```

## Example 2: pseudobulk pca + tscan + predefined 2 clusters connecting HD center and COVID center
```{r}
precluster = sample_info_pb_center$cluster; names(precluster) = sample_info_pb_center$Patient
sample_info2 = get_sample_pseudotime(x = t(pbpca[,1:2]), sample_metadata = patient_info, 
                                    sample_name = 'Patient', phenotype_name = 'Phenotype',
                                    method = 'tscan', cluster = precluster,
                                    reduce = F, startcluster = 1)
sample_info_pc2 = pbpca %>% 
  as.data.frame() %>%
  select(PC1:PC2) %>%
  mutate(Patient = rownames(.)) %>%
  inner_join(sample_info2 %>% select(Patient, Phenotype, cluster, Pseudotime), by = 'Patient')

create_pseudotime_plot(sample_info_pc2, "Gene Expression", cluster = T)
```

## Example 3: cell proportion pca + tscan + model-based clustering with 2 clusters
```{r}
sample_info3 = get_sample_pseudotime(x = t(cell_prop_pca[,1:2]), sample_metadata = patient_info, 
                                    sample_name = 'Patient', phenotype_name = 'Phenotype',
                                    method = 'tscan',
                                    reduce = F, clusternum = 2, startcluster = 1)
sample_info_pc3 = cell_prop_pca %>% 
  as.data.frame() %>%
  select(PC1:PC2) %>%
  mutate(Patient = rownames(.)) %>%
  inner_join(sample_info3 %>% select(Patient, Phenotype, cluster, Pseudotime), by = 'Patient')

create_pseudotime_plot(sample_info_pc3, "Cell Proportion", cluster = T)
```

## Example 4: pseudobulk pca + cca 
```{r}
sev.level = patient_info[match(rownames(pbpca),patient_info$Patient),]$sev.level
sample_info4 = get_sample_pseudotime(x = t(pbpca[,1:2]), sample_metadata = patient_info, 
                                     sample_name = 'Patient', phenotype_name = 'Phenotype',
                                     method = 'cca', phenotype_num = sev.level)
sample_info_pc4 = pbpca %>% 
  as.data.frame() %>%
  select(PC1:PC2) %>%
  mutate(Patient = rownames(.)) %>%
  inner_join(sample_info4 %>% select(Patient, Phenotype, Pseudotime), by = 'Patient')

create_pseudotime_plot(sample_info_pc4, "Gene Expression", cluster = F, opt_axis = T)

```


# Explore How Molecular Features Change Along Pseudotime
## Cell type proportion
We first load cell type proportion/count data and sample pseudotime results. 
```{r}
cell_type_prop = readRDS("./data/cell_type_prop.rds")
celltype_info = readRDS("./data/celltype_info.rds")
sample_info_opt = readRDS("./data/sample_info_pb_opt.rds")
```


Then we provide the following function `get_cell_prop_traj` to visualize the cell proportion change along pseudotime. 

* `cell_type_prop`: the cell type proportion data, each row is a cell cluster and each column is a sample.

* `cluster_list` and `celltype_list`: the cluster list and corresponding celltype list to visualize.

* `sample_metadata`: patient-level metadata, including several important variables.

* `sample_name` and `pseudotime_name`: the variable name in `sample_metadata` that indicates the sample and pseudotime. The default setting of `sample_name` is "Patient" and `phenotype_name` is "Pseudotime". 

```{r}
get_cell_prop_traj = function(cell_type_prop, cluster_list = NULL, celltype_list = NULL, 
                              sample_metadata = NULL, 
                              sample_name = 'Patient', pseudotime_name = 'Pseudotime'){
  celltype_label = paste0('c', cluster_list, '\n', celltype_list)
  patient_order = sample_metadata[, sample_name]
  cp = cell_type_prop[cluster_list, match(patient_order,colnames(cell_type_prop)),drop = F]
  
  cp_df = cp %>%
    as.data.frame() %>%
    mutate(celltype = celltype_label) %>%
    gather(Patient, value, -celltype) %>%
    inner_join(sample_metadata %>% select({{sample_name}}, {{pseudotime_name}}), by = setNames("Patient",sample_name))
  
  
  p = cp_df %>%
    mutate(celltype = factor(celltype, levels = celltype_label)) %>%
    ggplot(aes(x = Pseudotime, y = value))+
    geom_point()+ geom_smooth()+ 
    facet_wrap(vars(celltype),scales = "free_y")+
    ylab("Cell Type Proportion")+
    theme_classic()
  
  return(p)
    
}
```

Now we explore some important cell clusters in our COVID data
```{r}
topclu = c(1,2,3,4,8,13,14,24,25); names(topclu) = paste0("c",topclu)
topct = celltype_info$celltype[topclu]

p1 = get_cell_prop_traj(cell_type_prop,cluster_list = topclu[1], celltype_list = topct[1],
                       sample_metadata = sample_info_opt)
p1
p = get_cell_prop_traj(cell_type_prop,cluster_list = topclu, celltype_list = topct,
                       sample_metadata = sample_info_opt)
p
```

## Gene Expression
Now we will demonstrate how to detect differential genes along pseudotime, separately for each cell cluster, using the function `run_pseudo_diff_gene`

* `x`: the pseudobulk gene expression data, each row is a gene:cluster feature and each column is a sample.

* `cluster`: the cell cluster. 

* `sample_metadata`: patient-level metadata, including several important variables.

* `sample_name` and `pseudotime_name`: the variable name in `sample_metadata` that indicates the sample and pseudotime. The default setting of `sample_name` is "Patient" and `pseudotime_name` is "Pseudotime". 

* `plot`: logical. Indicate whether to plot the heatmap or not. 

```{r}
run_pseudo_diff_gene = function(x, sample_metadata = NULL, cluster = 1,
                         sample_name = 'Patient', pseudotime_name = 'Pseudotime',
                         plot = F
                         ){
  #print(cluster)
  ptime = sample_metadata[,pseudotime_name]
  sample_metadata = sample_metadata[order(ptime),]
  
  fn = rownames(x)
  gn = gsub('(.*):(.*)','\\1',fn)
  cn = gsub('(.*):(.*)','\\2',fn)
  
  xc = x[which(cn == paste0("c",cluster)),]
  xc_reorder = xc[,match(sample_metadata$Patient,colnames(x))]
  
  res = lapply(1:nrow(xc_reorder), function(i){
    model = mgcv::gam(xc_reorder[i,]~s(c(1:ncol(xc_reorder)),k=3))
    smooth = model$fitted.values
    pval = pchisq(model$null.deviance - model$deviance, model$df.null - model$df.residual,lower.tail = F) 
    list(smooth = smooth, pval = pval)
  })
  pval = sapply(res, function(i) i$pval)
  xsmooth = lapply(res, function(i) i$smooth) %>% do.call(rbind, .)
  rownames(xsmooth) = rownames(xc_reorder)
  
  
  fdr = p.adjust(pval, method = 'fdr')
  result = data.frame(gene = gsub('(.*):(.*)','\\1',rownames(xc_reorder)), pval = pval, fdr = fdr) %>%
    mutate(cluster = cluster,
           signif = (fdr < 0.05)) %>%
    arrange(fdr)
  print(result)
  sig = result %>% filter(signif) %>% pull(gene) 
  sig = paste0(sig,':c',cluster)
  
  xsig = xsmooth[rownames(xsmooth) %in% sig,]
  xsig = t(apply(xsig, 1, scale))
  rownames(xsig) = gsub('(.*):(.*)','\\1',rownames(xsig))
  colnames(xsig) = colnames(xsmooth)
  
  if (plot){
    Study.color = readRDS("./data/study.rds")
    Phenotype.color = readRDS("./data/severity.rds")
    patient_annot = HeatmapAnnotation(Severity = sample_metadata$Phenotype,
                                      Study = sample_metadata$Study,
                                      #Sex = sample_info$Sex,
                                      #Age = sample_info$Age,
                                      col = list(Severity = Phenotype.color,
                                                 Study = Study.color),
                                      #Sex = sex.color,
                                      #Age = col_fun),
                                      show_legend = F, simple_anno_size = unit(10,"mm"),
                                      annotation_name_gp = gpar(fontsize = 25)
    )
    p = Heatmap(xsig, name = "Expression", 
                show_row_dend = F, cluster_rows = T,
                column_title = "Pseudotime -->", column_title_side = "bottom",
                top_annotation = patient_annot, 
                cluster_columns = F, column_dend_reorder=F,
                row_names_gp = gpar(fontsize = 10), row_names_centered = T,
                column_names_gp = gpar(fontsize = 10),
                column_title_gp = gpar(fontsize = 40),
                show_heatmap_legend = F,
                show_column_names = F,
                heatmap_legend_param = list(
                  title_gp = gpar(fontsize = 25),
                  labels_gp = gpar(fontsize = 25),
                  legend_height = unit(8, "cm"),
                  grid_width = unit(1,"cm")
                ))
    print(p)
  }
  return(result)
}
```


```{r}
pb_example = readRDS("./data/pb_example_c1.rds")
common_pt = intersect(colnames(pb_example), sample_info_opt$Patient)
sample_info = sample_info_opt %>%
  filter(Patient %in% common_pt)

DEG_result = run_pseudo_diff_gene(x = pb_example, sample_metadata = sample_info, cluster = 1,
                           plot = T)
print(DEG_result)
```

We can also use the function `get_gene_expr_traj` to explore how the gene expression levels in certain clusters change over pseudotime. 


* `x`: the pseudobulk gene expression data, each row is a gene:cluster feature and each column is a sample.

* `gene_list`: the gene list to visualize. 

* `cluster`: the cell cluster. 

* `sample_metadata`: patient-level metadata, including several important variables.

* ``pseudotime_name`: the variable name in `sample_metadata` that indicates the pseudotime. The default setting of `pseudotime_name` is "Pseudotime". 


```{r}
get_gene_expr_traj = function(x, gene_list, cluster, sample_metadata, 
                              pseudotime_name = 'Pseudotime'){
  ptime = sample_metadata[,pseudotime_name]
  sample_metadata = sample_metadata[order(ptime),]
  xg = x[paste0(gene_list,":c",cluster),,drop = F]
  colnames(xg) = 1:ncol(xg)
  df = xg %>%
    as.data.frame() %>%
    mutate(gene=rownames(.)) %>%
    pivot_longer(!gene, names_to = 'Pseudotime', values_to = 'Expression') %>%
    mutate(Pseudotime = as.numeric(Pseudotime))
  str(df)
  p = df%>%
    ggplot(aes(x = Pseudotime, y = Expression))+
    geom_point()+ geom_smooth()+
    facet_wrap(vars(gene), scales = "free_y") + theme_classic()
  return(p)
}
p = get_gene_expr_traj(x = pb_example, 
                       gene_list = c('GZMA','RAP1GAP2','FXYD7','KLRB1','RPS27','EEF1A1'), 
                       cluster = 1, 
                       sample_metadata = sample_info)
p
```


# Session Info {.unnumbered}

```{r sessionInfo, echo=FALSE}
sessionInfo()
```
